src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs

```diff
@@ -132,102 +154,165 @@ public virtual byte[] GetValue()
-                        attribute = (string)o[1];
+                        PageResultResponseControl pageControl = new PageResultResponseControl(size, cookie, controls[i].IsCritical, value);
+                        controls[i] = pageControl;
                     }
-                    else
+                    else if (controls[i].Type == "1.2.840.113556.1.4.1504")
                     {
-                        // decoding might fail as attribute is optional
-                        o = BerConverter.Decode("{e}", value);
-                        Debug.Assert(o != null && o.Length == 1);
+                        // The control is an AsqControl, as described in MS-ADTS section 3.1.1.3.4.1.18
+                        AsnDecoder.ReadSequence(asnSpan, AsnEncodingRules.BER, out int sequenceContentOffset, out int sequenceContentLength, out _);
+                        ThrowUnless(sequenceContentLength > 0);
 
-                        result = (int)o[0];
-                    }
+                        ResultCode result = AsnDecoder.ReadEnumeratedValue<ResultCode>(asnSpan.Slice(sequenceContentOffset, sequenceContentLength), AsnEncodingRules.BER, out _);
 
-                    SortResponseControl sort = new SortResponseControl((ResultCode)result, attribute, controls[i].IsCritical, controls[i].GetValue());
-                    controls[i] = sort;
-                }
-                else if (controls[i].Type == "2.16.840.1.113730.3.4.10")
-                {
-                    // The control is a VlvResponseControl.
-                    int position;
-                    int count;
-                    int result;
-                    byte[] context = null;
-                    object[] o = BerConverter.TryDecode("{iieO}", value, out bool decodeSucceeded);
-
-                    if (decodeSucceeded)
+                        AsqResponseControl asq = new AsqResponseControl(result, controls[i].IsCritical, value);
+                        controls[i] = asq;
+                    }
+                    else if (controls[i].Type == "1.2.840.113556.1.4.841")
                     {
-                        Debug.Assert(o != null && o.Length == 4);
-                        position = (int)o[0];
-                        count = (int)o[1];
-                        result = (int)o[2];
-                        context = (byte[])o[3];
+                        // The control is a DirSyncControl, as described in MS-ADTS section 3.1.1.3.4.1.3
+                        byte[] dirsyncCookie;
+
+                        AsnDecoder.ReadSequence(asnSpan, AsnEncodingRules.BER, out int sequenceContentOffset, out int sequenceContentLength, out _);
+                        ThrowUnless(sequenceContentLength > 0);
+                        asnSpan = asnSpan.Slice(sequenceContentOffset, sequenceContentLength);
+
+                        asnReadSuccessful = AsnDecoder.TryReadInt32(asnSpan, AsnEncodingRules.BER, out int moreData, out int bytesConsumed);
```

Comment:
I believe `moreData` here corresponds to `Flags` from the DirSyncRequestValue structure, and should probably be named similarly.

Reply:
Technically, nothing constrains `Flags` to be System.Int32; but it looks like BerConverter.Decode `i` is treated as System.Int32; so this is just maintaining the limits that the previous version had.

If it just happened once or twice I'd give a comment saying why the code thought 32-bit was sufficient; but since it's pervasive... I'd consider putting a descriptive comment at the top, like

```
// Throughout this code, structure reads treat BER INTEGER values as System.Int32.
// Previous versions of this class, including the version in .NET Framework,
// used BerConverter.Decode with 'i' for the integer types, and that data type code is
// mapped to System.Int32.
// Therefore, when using AsnDecoder we have historical precedent to justify limiting to 32-bit.
```

Reply:
I've added the comments; thanks.
This control in specific corresponds to `DirSyncResponseValue`. `moreData` corresponds to the `MoreResults` field in that structure. I've renamed the variable to match it more closely though.